### PR TITLE
Enriched egress events from flyteadmin

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR
Allow flyteadmin to publish cloud events to the new topic

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
execution_manager publish a cloudevent to the new topic when invoking [CreateWorkflowEvent](https://github.com/flyteorg/flyteadmin/blob/412b834a76c98edfe95bc82eda6a7d23da504c9a/pkg/manager/impl/execution_manager.go#L1162)

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1019

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/async/notifications/implementations/cloudevent_publisher.go
Comment (wild-endeavor) [inline pkg/async/notifications/implementations/cloudevent_publisher.go:23]: i think for the initial implementation we only need one or two events, but I think the structure should be there to be able to translate all events we have - all of these https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/event/event.proto

Where is the filter currently to decide which events to translate from raw event to cloudevent?

**Review by wild-endeavor — COMMENTED** [review]
I haven't looked too closely at the CloudEvent spec.  It doesn't feel like it does a whole lot right now.  It feels like we're just doing event.SetData... which is fine but it doesn't really buy us that much.  I was hoping there would at least be a json schema or something we could publish as part of the cloudevent spec.

does this make sense?  the broader goal of going through cloudevent was that we could transform the raw flyte events into a standardized set of even information that wouldn't change even if the backend raw events changed.

File: Dockerfile
Comment (pingsutw) [reply Dockerfile:14]: The error occurred every time I built the image `make docker_build`, so I added it. It only happened when I changed the code

anyway, will remove it before merging.

```bash
#13 0.249 	k8s.io/utils@v0.0.0-20210305010621-2afb4311ab10: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
#13 0.249 	sigs.k8s.io/structured-merge-diff/v4@v4.1.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
#13 0.249 	sigs.k8s.io/yaml@v1.2.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
#13 0.249 	github.com/robfig/cron/v3: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
#13 0.249 
#13 0.249 	To ignore the vendor directory, use -mod=readonly or -mod=mod.
#13 0.249 	To sync the vendor directory, run:
#13 0.249 		go mod vendor
#13 0.250 make: *** [Makefile:46: linux_compile] Error 1
------
executor failed running [/bin/sh -c make linux_compile]: exit code: 2
make: *** [docker_build] Error 1
```

File: pkg/async/notifications/implementations/cloudevent_publisher.go
Comment (pingsutw) [reply pkg/async/notifications/implementations/cloudevent_publisher.go:23]: > Where is the filter currently to decide which events to translate from raw event to cloudevent?

here. https://github.com/flyteorg/flyteadmin/pull/355/files#diff-899da2dc4f47b8d6467d0d2a8d779f36fcb7334965153ceae018e2652e9f0ac7R36-R38

File: Dockerfile
Comment (wild-endeavor) [reply Dockerfile:14]: that's weird.  @EngHabu do you have any ideas?

File: pkg/async/notifications/implementations/cloudevent_publisher.go
Comment (katrogan) [inline pkg/async/notifications/implementations/cloudevent_publisher.go:92]: is this final? are we checking this into idl?

File: pkg/async/notifications/implementations/cloudevent_publisher.go
Comment (katrogan) [inline pkg/async/notifications/implementations/cloudevent_publisher.go:124]: panic here?

**Review by katrogan — COMMENTED** [review]
top-level questions

- do we want calls to publish need to be async so that they don't block the execution event write path? is lossiness okay?
- for clarity, instead of adding this code in notifications/ can we create a separate folder under async/

File: pkg/async/notifications/implementations/cloudevent_publisher.go
Comment (pingsutw) [reply pkg/async/notifications/implementations/cloudevent_publisher.go:79]: okay.
Should we also move it below the call to `!p.shouldPublishEvent`  in `event_publisher`?
https://github.com/flyteorg/flyteadmin/blob/c80c8d8fd97f607da83db3ebb84cf0aec3b1f191/pkg/async/notifications/implementations/event_publisher.go#L53

File: pkg/async/notifications/implementations/cloudevent_publisher.go
Comment (pingsutw) [reply pkg/async/notifications/implementations/cloudevent_publisher.go]: sorry, the link is outdated. The new schema url is https://github.com/flyteorg/flyteidl/blob/cloudevent2/jsonschema/workflow_execution.json

**Comment by pingsutw** [comment]
> do we want calls to publish need to be async so that they don't block the execution event write path? is lossiness okay?

I've updated the code. To minimize overhead, use goroutine to publish cloudevent instead.

> for clarity, instead of adding this code in notifications/ can we create a separate folder under async/

Updated it.

File: pkg/async/cloudevent/cloudevent_publisher.go
Comment (katrogan) [inline pkg/async/cloudevent/cloudevent_publisher.go:72]: why not re-use the input context?

File: pkg/async/cloudevent/cloudevent_publisher.go
Comment (katrogan) [inline pkg/async/cloudevent/cloudevent_publisher.go:104]: nit: include the original msg here too?

File: pkg/async/notifications/factory.go
Comment (katrogan) [inline pkg/async/notifications/factory.go:240]: can we move this factory method under the cloudevent dir too?

File: pkg/async/notifications/factory.go
Comment (katrogan) [inline pkg/async/notifications/factory.go:288]: let's pass in a context into this func and use that here

File: pkg/runtime/application_config_provider.go
Comment (katrogan) [inline pkg/runtime/application_config_provider.go:91]: can we add sensible defaults for `ReconnectAttempts` and `ReconnectDelaySeconds`?

File: pkg/runtime/interfaces/application_configuration.go
Comment (katrogan) [inline pkg/runtime/interfaces/application_configuration.go:450]: should this move to the KafkaConfig?

**Review by katrogan — COMMENTED** [review]
this is looking great! thanks for adding the comprehensive unit tests

File: pkg/async/cloudevent/cloudevent_publisher.go
Comment (pingsutw) [reply pkg/async/cloudevent/cloudevent_publisher.go:72]: updated it, sorry. my bad

File: pkg/async/cloudevent/cloudevent_publisher.go
Comment (pingsutw) [reply pkg/async/cloudevent/cloudevent_publisher.go]: Updated it

File: pkg/async/cloudevent/implementations/cloudevent_publisher_test.go
Comment (pingsutw) [reply pkg/async/cloudevent/implementations/cloudevent_publisher_test.go:25]: sorry, typo

File: pkg/async/cloudevent/implementations/sender.go
Comment (pingsutw) [reply pkg/async/cloudevent/implementations/sender.go:57]: I think we don't need it. There is a retry mechanism in cloudevent SDK, so we don't need to handle it.

File: pkg/async/cloudevent/implementations/sender.go
Comment (pingsutw) [reply pkg/async/cloudevent/implementations/sender.go:22]: updated it

File: pkg/async/cloudevent/factory.go
Comment (pingsutw) [reply pkg/async/cloudevent/factory.go:23]: The interfaces are the same now, but I do agree to have a new interface for cloudevent publisher. 
Yeah, we may extend `cloudevent.publisher` in the future.

File: pkg/async/cloudevent/implementations/sender.go
Comment (pingsutw) [reply pkg/async/cloudevent/implementations/sender.go]: The new file structure will be like
```
cloudevent
├── factory.go
├── factory_test.go
├── implementations
│   ├── cloudevent_publisher.go
│   ├── cloudevent_publisher_test.go
│   ├── sender.go
│   └── sender_test.go
├── interfaces
│   ├── publisher.go
│   └── sender.go
└── mocks
    ├── publisher.go
    └── sender.go
```

File: pkg/async/cloudevent/implementations/sender.go
Comment (katrogan) [reply pkg/async/cloudevent/implementations/sender.go:51]: awesome

File: pkg/async/cloudevent/implementations/sender.go
Comment (katrogan) [reply pkg/async/cloudevent/implementations/sender.go]: great, thank you!

File: pkg/runtime/application_config_provider.go
Comment (katrogan) [reply pkg/runtime/application_config_provider.go:91]: ah okay, staying consistent is fine too

**Review by katrogan — DISMISSED** [review]

**Review by katrogan — APPROVED** [review]

File: pkg/async/cloudevent/implementations/cloudevent_publisher.go
Comment (katrogan) [inline pkg/async/cloudevent/implementations/cloudevent_publisher.go:28]: will you check this into master? can you hold off on merging this change until you do?

Comment (pingsutw) [reply pkg/async/cloudevent/implementations/cloudevent_publisher.go:28]: yeah, we should merge this PR https://github.com/flyteorg/flyteidl/pull/270 first.  btw, will fix the test errors. cc @wild-endeavor

**Review by katrogan — APPROVED** [review]

**Pull request merged**
